tools/xdp/s2n-quic-xdp/src/io/rx.rs

```diff
@@ -0,0 +1,167 @@
+        ready!(occupied)?;
+
+        Poll::Ready(Ok(()))
+    }
+
+    #[inline]
+    fn queue<F: FnOnce(&mut Self::Queue)>(&mut self, f: F) {
+        let this: &'static mut Self = unsafe {
+            // Safety: As noted in the [transmute examples](https://doc.rust-lang.org/std/mem/fn.transmute.html#examples)
+            // it can be used to temporarily extend the lifetime of a reference. In this case, we
+            // don't want to use GATs until the MSRV is >=1.65.0, which means `Self::Queue` is not
+            // allowed to take generic lifetimes.
+            //
+            // We are left with using a `'static` lifetime here and encapsulating it in a private
+            // field. The `Self::Queue` struct is then borrowed for the lifetime of the `F`
+            // function. This will prevent the value from escaping beyond the lifetime of `&mut
+            // self`.
+            //
+            // See https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9a32abe85c666f36fb2ec86496cc41b4
+            //
+            // Once https://github.com/aws/s2n-quic/issues/1742 is resolved this code can go away
+            core::mem::transmute(self)
+        };
+
+        let occupied = this.occupied.slice();
+        let free = this.free.slice();
+        let umem = &mut this.umem;
+        let error_logger = &mut this.error_logger;
+
+        let mut queue = Queue {
+            occupied,
+            free,
+            umem,
+            error_logger,
+        };
+        f(&mut queue);
+    }
+
+    #[inline]
+    fn handle_error<E: event::EndpointPublisher>(self, _error: Self::Error, _events: &mut E) {
+        // no-op, just shut down
+    }
+}
+
+pub struct Queue<'a> {
+    occupied: spsc::RecvSlice<'a, RxTxDescriptor>,
+    free: spsc::SendSlice<'a, UmemDescriptor>,
+    umem: &'a mut Umem,
+    error_logger: &'a mut Option<Box<dyn ErrorLogger>>,
+}
+
+impl<'a> rx::Queue for Queue<'a> {
+    type Handle = path::Tuple;
+
+    #[inline]
+    fn for_each<F: FnMut(datagram::Header<Self::Handle>, &mut [u8])>(&mut self, mut on_packet: F) {
+        // only pop as many items as we have capacity to free them
+        while self.free.capacity() > 0 {
+            let descriptor = match self.occupied.pop() {
+                Some(v) => v,
+                None => break,
+            };
+
+            let buffer = unsafe {
+                // Safety: this descriptor should be unique, assuming the tasks are functioning
+                // properly
+                self.umem.get_mut(descriptor)
+            };
+
+            // create a decoder from the descriptor's buffer
+            let decoder = DecoderBufferMut::new(buffer);
+
+            // try to decode the packet and emit the result
+            match decoder::decode_packet(decoder) {
+                Ok(Some((header, payload))) => {
+                    on_packet(header, payload.into_less_safe_slice());
+                }
+                Ok(None) | Err(_) => {
```

Comment:
are the contents of `Err` useful to log?

Reply:
I don't think so... It would just be a decoder error with something like "unexpected end of buffer" which isn't super useful. In either case, this function being called likely means that nothing will get through to the quic endpoint.